### PR TITLE
Azure Cloud Service Tracing Improvements

### PR DESCRIPTION
**Pull request by eisber**

### CODE REVIEW

**Comment by JohnLangford** [comment]
@lhoang29 can you take a look at this?

**Comment by lhoang29** [comment]
looks good

**Comment by JohnLangford** [comment]
Travis is failing on this one.

**Comment by eisber** [comment]
hmm some of my unfinished LDA improvements made into this. can you hold of a bit?

**Comment by JohnLangford** [comment]
Status on this?

**Comment by lhoang29** [comment]
Do I need to review more?

**Comment by eisber** [comment]
fixed Dictionary<string, float> support
fixed Test 77

**Comment by JohnLangford** [comment]
@lhoang29 Can you take a look again?

File: cs/azure/Trainer/Operations/InitializeOperation.cs [hunk 1]
```diff
@@ -75,7 +75,11 @@ internal void FreshStart(OnlineTrainerState state = null)
             // start from scratch
             this.state = state;
 
-            this.InitializeVowpalWabbit(new VowpalWabbitSettings(this.settings.Metadata.TrainArguments));
+            var settings = model == null ?
+                new VowpalWabbitSettings(this.settings.Metadata.TrainArguments) :
+                new VowpalWabbitSettings(string.Empty) { ModelStream = new MemoryStream(model) };
```
Comment (lhoang29) [inline cs/azure/Trainer/Operations/InitializeOperation.cs]: This is probably wrapped in a try-catch under the PUT method right? Also, we'll likely see many user errors w.r.t. using warm start model so perhaps in the logged message, we could have a link to some wiki page for diagnosing it? By wiki page I'm thinking of general guidelines on what kind of models would be compatible etc...

File: cs/azure/Worker/OnlineTrainerSettingsWatcher.cs [hunk 2]
```diff
@@ -152,7 +152,7 @@ private ICheckpointPolicy ParseCheckpointPolicy()
                 if (checkpointString.Contains(":"))
                 {
                     TimeSpan interval;
-                    if (!TimeSpan.TryParse(checkpointString, CultureInfo.InvariantCulture, out interval))
+                    if (TimeSpan.TryParse(checkpointString, CultureInfo.InvariantCulture, out interval))
```
Comment (lhoang29) [inline cs/azure/Worker/OnlineTrainerSettingsWatcher.cs]: this seems useful :)

File: cs/azure/Worker/ResetController.cs [hunk 3]
```diff
@@ -57,5 +66,29 @@ public ResetController(LearnEventProcessorHost trainProcessorFactory)
...
+
+            try
+            {
+                var model = await Request.Content.ReadAsByteArrayAsync();
+
+                await this.trainProcessorHost.ResetModelAsync(model: model);
+
+                return this.Request.CreateResponse(HttpStatusCode.OK);
+            }
+            catch (Exception ex)
```
Comment (lhoang29) [inline cs/azure/Worker/ResetController.cs]: ah this answers part of my other question

File: explore/cpp/utility.h [hunk 4]
```diff
@@ -232,7 +232,7 @@ class NumberUtils
 class StringUtils
 {
 public:
-  static inline string to_string(size_t& n)
+  static inline string to_string(size_t n)
```
Comment (lhoang29) [inline explore/cpp/utility.h]: why is this needed?

Comment (eisber) [inline explore/cpp/utility.h]: In cbify.cc I’m calling to_string(<temp result>) and one can’t take &<temp result>

Also I don’t see any benefit of using the reference here…

File: vowpalwabbit/lda_core.cc [hunk 5]
```diff
@@ -809,12 +858,227 @@ void learn(lda &l, LEARNER::base_learner &, example &ec)
...
+struct feature_pair
+{
+	uint64_t f1;
+	uint64_t f2;
+
+	feature_pair(uint64_t _f1, uint64_t _f2) : f1(_f1), f2(_f2)
+	{}
+};
+
+void compute_metrics(lda &l)
```
Comment (lhoang29) [inline vowpalwabbit/lda_core.cc]: Perhaps add comments / clean up this part to make it more readable. Additionally, should this be named something like compute_coherence?

File: explore/cpp/utility.h (hunk 4 above)
Comment (lhoang29) [inline explore/cpp/utility.h]: Doesn't look like this is on the hot path at the moment so not a real issue. Just seemed odd to pass by value but I see this avoids an awkward new variable.

File: vowpalwabbit/cbify.cc [hunk 6]
```diff
@@ -40,6 +40,8 @@ struct cbify
   MwtExplorer<example>* mwt_explorer;
   vw_recorder* recorder;
   v_array<action_score> a_s;
+  // used as the seed
+  size_t example_counter;
```
Comment (lhoang29) [inline vowpalwabbit/cbify.cc]: Is this being initialized somewhere? I see it's just being used as a seed so bogus starting values are still ok, although someone else could use it for its name.

**Comment by lhoang29** [comment]
done with another pass

File: vowpalwabbit/cbify.cc (hunk 6 above)
Comment (eisber) [inline vowpalwabbit/cbify.cc]: I joined the dark side by relying on calloc. John laughed pretty hard on me when he saw me doing it…

File: cs/azure/Trainer/Operations/InitializeOperation.cs (hunk 1 above)
Comment (eisber) [inline cs/azure/Trainer/Operations/InitializeOperation.cs]: done

**Comment by JohnLangford** [comment]
Status on this?

**Comment by eisber** [comment]
Ready to pull…

**Comment by JohnLangford** [comment]
Merged, thanks.
